File: Learning/src/StackExample.java
import java.util.Stack;

public class StackExample {
    public static void main(String[] args) {
        // Signature
        Stack<Integer> stack = new Stack<>();

        // Push
        System.out.println("Pushing element onto the stack: 1");
        stack.push(1);

        // Pop
        if (!stack.isEmpty()) {
            int poppedElement = stack.pop();
            System.out.println("Popped element from the stack: " + poppedElement);
        } else {
            System.out.println("Stack is empty. Cannot pop.");
        }

        // Peek
        if (!stack.isEmpty()) {
            int topElement = stack.peek();
            System.out.println("Top element of the stack: " + topElement);
        } else {
            System.out.println("Stack is empty. No top element to peek.");
        }

        // Add more elements
        stack.push(2);
        stack.push(3);
        System.out.println("Stack size: " + stack.size());

        // Iterate through Stack using for-each loop
        System.out.println("\nIterating through Stack using for-each loop:");
        for (int element : stack) {
            System.out.println("Element: " + element);
        }

        // Alternatively, use Iterator for more control
        System.out.println("\nIterating using Iterator:");
        java.util.Iterator<Integer> iterator = stack.iterator();
        while (iterator.hasNext()) {
            int element = iterator.next();
            System.out.println("Element: " + element);
        }
        
        // Alternatively, use forEach method (Java 8 and later)
        System.out.println("\nIterating using forEach method:");
        stack.forEach(element -> System.out.println("Element: " + element));

        // Alternatively, convert to array and use enhanced for loop
        System.out.println("\nIterating using toArray method:");
        Integer[] elementsArray = stack.toArray(new Integer[0]);
        for (int element : elementsArray) {
            System.out.println("Element: " + element);
        }
    }
}

File: DSA/Python/DataStructures/stack_example.py
def main():
    # Initialize stack
    stack = []

    # Push
    print("Pushing element onto the stack: 1")
    stack.append(1)

    # Pop
    if stack:
        popped_element = stack.pop()
        print(f"Popped element from the stack: {popped_element}")
    else:
        print("Stack is empty. Cannot pop.")

    # Peek
    if stack:
        top_element = stack[-1]
        print(f"Top element of the stack: {top_element}")
    else:
        print("Stack is empty. No top element to peek.")

    # Add more elements
    stack.append(2)
    stack.append(3)
    print(f"Stack size: {len(stack)}")

    # Different ways to iterate
    print("\nIterating through Stack using for loop:")
    for element in stack:
        print(f"Element: {element}")

    print("\nIterating using enumerate:")
    for i, element in enumerate(stack):
        print(f"Element: {element}")

    print("\nIterating using list comprehension:")
    [print(f"Element: {element}") for element in stack]

if __name__ == "__main__":
    main()
